### PR TITLE
Avoid routing through AMPL on second hop

### PR DESCRIPTION
**Pull request by mikeki**
- **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Feature/BugFix

- **What is the current behavior?** (You can also link to an open issue here)
Currently AMPL, being a rebasing token, is paired with other tokens (namely COMP) in a way that the pools are broken. The AMPL/COMP pool is not synced, so the quoter believes theres more liquidity than there really is.

- **What is the new behavior (if this is a feature change)?**
There's not an easy solution to this problem, so the approach we are taking is of ignoring all AMPL paired pools on second hops.
This feature actually supports adding more tokens to this list as needed. 

- **Other information**:

### CODE REVIEW

**Review by cgkol — COMMENTED** [review]
Lets please add a test for this!

**Comment by jsy1218** [comment]
I'm a bit worried about the tenderly integ-test failures:

```
  ● alpha router integration › mainnet alpha - 0 › + Simulate on Tenderly + Execute on Hardhat fork › DAI -> WBTC with portion

    expect(received).toEqual(expected) // deep equality

    Expected: 2
    Received: 1

      2448 |
      2449 |               // Expect tenderly simulation to be successful
    > 2450 |               expect(swap!.simulationStatus).toEqual(SimulationStatus.Succeeded);
           |                                              ^
      2451 |               expect(swap!.methodParameters).toBeDefined();
      2452 |               expect(swap!.methodParameters!.to).toBeDefined();
```

1 means `SimulationStatus.Failed`. Worth looking into it, before we merge the PR.

File: src/routers/alpha-router/functions/get-candidate-pools.ts [hunk 1]
```diff
@@ -899,11 +903,25 @@ export async function getV2CandidatePools({
...
+  }).map((pool) =>
+    tokenInAddress === pool.token0.id ? pool.token1.id : pool.token0.id
   );
-  const tokenOutSecondHopAddresses = topByTVLUsingTokenOut.map((pool) =>
-    tokenOutAddress == pool.token0.id ? pool.token1.id : pool.token0.id
+  const tokenOutSecondHopAddresses = topByTVLUsingTokenOut.filter((pool) => {
+    // filtering second hops
+    if (tokenInAddress === pool.token0.id) {
+      return !tokensToAvoidOnSecondHops?.includes(pool.token1.id.toLowerCase());
+    } else {
```
Comment (jsy1218) [inline src/routers/alpha-router/functions/get-candidate-pools.ts:920]: do you need else if here:

```
if (tokenInAddress === pool.token0.id) {
   return !tokensToAvoidOnSecondHops?.includes(pool.token1.id.toLowerCase());
} else if (tokenInAddress === pool.token1.id) {
   return !tokensToAvoidOnSecondHops?.includes(pool.token0.id.toLowerCase());
}
```

File: src/routers/alpha-router/functions/get-candidate-pools.ts [hunk 2]
```diff
@@ -899,11 +903,25 @@ export async function getV2CandidatePools({
   // Filtering step for second hops
   const topByTVLUsingTokenInSecondHopsMap: Map<string, SubcategorySelectionPools<V2SubgraphPool>> = new Map();
   const topByTVLUsingTokenOutSecondHopsMap: Map<string, SubcategorySelectionPools<V2SubgraphPool>> = new Map();
-  const tokenInSecondHopAddresses = topByTVLUsingTokenIn.map((pool) =>
-    tokenInAddress == pool.token0.id ? pool.token1.id : pool.token0.id
+  const tokenInSecondHopAddresses = topByTVLUsingTokenIn.filter((pool) => {
+    // filtering second hops
+    if (tokenInAddress === pool.token0.id) {
+      return !tokensToAvoidOnSecondHops?.includes(pool.token1.id.toLowerCase());
+    } else {
```
Comment (jsy1218) [inline src/routers/alpha-router/functions/get-candidate-pools.ts:910]: likewise:

```
if (tokenInAddress === pool.token0.id) {
   return !tokensToAvoidOnSecondHops?.includes(pool.token1.id.toLowerCase());
} else if (tokenInAddress === pool.token1.id) {
   return !tokensToAvoidOnSecondHops?.includes(pool.token0.id.toLowerCase());
}
```

**Comment by jsy1218** [comment]
> I'm a bit worried about the tenderly integ-test failures:
> 
> ```
>   ● alpha router integration › mainnet alpha - 0 › + Simulate on Tenderly + Execute on Hardhat fork › DAI -> WBTC with portion
> 
>     expect(received).toEqual(expected) // deep equality
> 
>     Expected: 2
>     Received: 1
> 
>       2448 |
>       2449 |               // Expect tenderly simulation to be successful
>     > 2450 |               expect(swap!.simulationStatus).toEqual(SimulationStatus.Succeeded);
>            |                                              ^
>       2451 |               expect(swap!.methodParameters).toBeDefined();
>       2452 |               expect(swap!.methodParameters!.to).toBeDefined();
> ```
> 
> 1 means `SimulationStatus.Failed`. Worth looking into it, before we merge the PR.

This is no longer a concern, now that https://github.com/Uniswap/smart-order-router/pull/436 fixed all the tenderly simulation failures.

File: src/routers/alpha-router/functions/get-candidate-pools.ts (hunk 1 above)
Comment (mikeki) [reply src/routers/alpha-router/functions/get-candidate-pools.ts:920]: I think it is safe to assume that tokenInAddress is token1 in that case.

**Review by jsy1218 — APPROVED** [review]

**Pull request merged**